File: backend/app/models/audit.py
"""Immutable audit log for sensitive actions."""
from __future__ import annotations

from sqlalchemy import ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base
from app.models.base import TimestampMixin


class AuditLog(Base, TimestampMixin):
    __tablename__ = "audit_logs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True, index=True)
    user_email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    organization_id: Mapped[int | None] = mapped_column(ForeignKey("organizations.id"), nullable=True, index=True)
    action: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    entity: Mapped[str | None] = mapped_column(String(100), nullable=True)
    entity_id: Mapped[str | None] = mapped_column(String(80), nullable=True)
    before: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    after: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    ip_address: Mapped[str | None] = mapped_column(String(64), nullable=True)
    user_agent: Mapped[str | None] = mapped_column(Text, nullable=True)

File: bluehost/public/app/Audit.php
<?php
// Append-only audit log writer.

class Audit
{
    public static function record(string $action, ?array $user = null, array $opts = []): void
    {
        Database::insert('audit_logs', [
            'user_id'       => $user['id'] ?? null,
            'user_email'    => $user['email'] ?? null,
            'organization_id' => $opts['organization_id'] ?? null,
            'action'        => $action,
            'entity'        => $opts['entity'] ?? null,
            'entity_id'     => isset($opts['entity_id']) ? (string) $opts['entity_id'] : null,
            'before_json'   => isset($opts['before']) ? json_encode($opts['before']) : null,
            'after_json'    => isset($opts['after']) ? json_encode($opts['after']) : null,
            'ip_address'    => $_SERVER['REMOTE_ADDR'] ?? null,
            'user_agent'    => $_SERVER['HTTP_USER_AGENT'] ?? null,
        ]);
    }
}
